File: examples/10-queues/java/src/main/java/io/orionmesh/queue/Processor.java
package io.orionmesh.queue;

import io.nats.client.*;
import io.nats.client.api.*;

import java.time.Duration;
import java.util.List;

/**
 * OrionMesh queue processor — Java reference.
 *
 * Reads configuration from environment variables injected by the Service spec
 * that {@code orion gen processor --lang java ...} produces:
 *
 * <pre>
 *   ORION_QUEUE_NAME    queue name (display only)
 *   ORION_QUEUE_SUBJECT subject to filter on
 *   ORION_QUEUE_STREAM  JetStream stream name
 *   ORION_QUEUE_TYPE    "work" or "topic"
 *   ORION_QUEUE_GROUP   durable consumer name (shared = balanced; unique = broadcast)
 *   ORION_REPLICA_INDEX agent-injected replica id
 *   NATS_URL            broker URL
 * </pre>
 *
 * Replace {@link #handle(String, String)} with your per-row logic.
 *
 * Debug attach (when started via {@code orion gen processor --lang java --debug ...}):
 *   1. {@code orion logs Service <name>} — wait for "Listening for transport dt_socket at address: 5005"
 *   2. In IntelliJ: Run > Edit Configurations > "+" > Remote JVM Debug, host=localhost port=5005
 *   3. Set a breakpoint in {@link #handle} — message arrival hits it.
 */
public class Processor {
    static final String NATS_URL    = env("NATS_URL", "nats://127.0.0.1:4222");
    static final String QUEUE_NAME  = env("ORION_QUEUE_NAME", "unnamed");
    static final String SUBJECT     = env("ORION_QUEUE_SUBJECT", "orion.queue." + QUEUE_NAME);
    static final String STREAM      = env("ORION_QUEUE_STREAM", "ORION_QUEUE_" + QUEUE_NAME.toUpperCase().replace('-', '_'));
    static final String QTYPE       = env("ORION_QUEUE_TYPE", "work");
    static final String BASE_GROUP  = env("ORION_QUEUE_GROUP", QUEUE_NAME + "-workers");
    static final String REPLICA     = env("ORION_REPLICA_INDEX", "0");
    // topic queues need a per-replica durable so JetStream tracks an
    // independent cursor (broadcast); work queues share the base group so
    // JetStream load-balances messages.
    static final String GROUP       = "work".equals(QTYPE) ? BASE_GROUP : BASE_GROUP + "-r" + REPLICA;
    static final String LABEL       = QUEUE_NAME + "#r" + REPLICA;

    /** Replace with your business logic. Each message body is a UTF-8 ndjson line. */
    static void handle(String subject, String payload) {
        System.out.printf("[%s] processed %s: %s%n", LABEL, subject,
                payload.length() > 200 ? payload.substring(0, 200) + "..." : payload);
    }

    public static void main(String[] args) throws Exception {
        System.out.printf("[%s] starting — type=%s subject=%s stream=%s group=%s%n",
                LABEL, QTYPE, SUBJECT, STREAM, GROUP);

        Options opts = new Options.Builder().server(NATS_URL).build();
        try (Connection nc = Nats.connect(opts)) {
            JetStream js = nc.jetStream();
            JetStreamManagement jsm = nc.jetStreamManagement();

            // Ensure the stream exists. Idempotent — `orion queue pub` does the same.
            try {
                jsm.getStreamInfo(STREAM);
            } catch (JetStreamApiException e) {
                StreamConfiguration sc = StreamConfiguration.builder()
                        .name(STREAM).addSubjects(SUBJECT).build();
                jsm.addStream(sc);
                System.out.printf("[%s] created stream %s%n", LABEL, STREAM);
            }

            // Idempotent ensure-consumer.
            ConsumerConfiguration cc = ConsumerConfiguration.builder()
                    .durable(GROUP)
                    .filterSubject(SUBJECT)
                    .ackPolicy(AckPolicy.Explicit)
                    .build();
            try {
                jsm.addOrUpdateConsumer(STREAM, cc);
            } catch (JetStreamApiException e) {
                System.err.printf("[%s] could not create consumer: %s%n", LABEL, e.getMessage());
                throw e;
            }

            PullSubscribeOptions pso = PullSubscribeOptions.builder().durable(GROUP).stream(STREAM).build();
            JetStreamSubscription sub = js.subscribe(SUBJECT, pso);
            System.out.printf("[%s] bound to durable=%s%n", LABEL, GROUP);

            while (true) {
                List<Message> msgs = sub.fetch(1, Duration.ofSeconds(10));
                for (Message m : msgs) {
                    try {
                        handle(m.getSubject(), new String(m.getData()));
                        m.ack();
                    } catch (RuntimeException ex) {
                        System.err.printf("[%s] handler error: %s — naking%n", LABEL, ex.getMessage());
                        m.nak();
                    }
                }
            }
        }
    }

    private static String env(String k, String d) {
        String v = System.getenv(k);
        return (v == null || v.isEmpty()) ? d : v;
    }
}

File: examples/10-queues/python/processor.py
#!/usr/bin/env python3
"""OrionMesh queue processor — Python reference.

Reads its configuration from environment variables injected by the Service
spec that `orion gen processor` produces:

    ORION_QUEUE_NAME    queue name (display only)
    ORION_QUEUE_SUBJECT subject to filter on
    ORION_QUEUE_STREAM  JetStream stream name (must already exist; the agent
                        ensures it via `orion queue pub` or `orion apply`)
    ORION_QUEUE_TYPE    "work" or "topic"
    ORION_QUEUE_GROUP   durable consumer name (shared for work, unique for topic)
    ORION_REPLICA_INDEX agent-injected replica id (0..replicas-1)
    NATS_URL            broker URL (default nats://127.0.0.1:4222)

Replace `handle(row)` with your own logic. Each row is a Python dict — the
parsed ndjson message.

Debug attach (when started via `orion gen processor --debug ...`):
    1. `orion logs Service <name>` — wait for "debugpy listening on 0.0.0.0:5678"
    2. In VS Code: Run > Add Configuration > Python: Remote Attach
                   host: localhost, port: 5678
    3. Set a breakpoint in handle() — message arrival hits it.
"""
from __future__ import annotations

import asyncio
import json
import os
import sys

try:
    import nats
    from nats.js.api import ConsumerConfig, AckPolicy, StreamConfig
    from nats.js.errors import NotFoundError
except ImportError:
    sys.stderr.write("nats-py not installed. Run: pip install -r requirements.txt\n")
    sys.exit(1)


def handle(row: dict) -> None:
    """User-editable per-row handler. Replace with your own logic.

    Set a breakpoint here when attaching with a debugger.
    """
    print(
        f"[{LABEL}] processed {row.get('_subject', '?')}: "
        f"{json.dumps(row, sort_keys=True)[:200]}",
        flush=True,
    )


NATS_URL = os.environ.get("NATS_URL", "nats://127.0.0.1:4222")
QUEUE_NAME = os.environ.get("ORION_QUEUE_NAME", "unnamed")
SUBJECT = os.environ.get("ORION_QUEUE_SUBJECT", f"orion.queue.{QUEUE_NAME}")
STREAM = os.environ.get("ORION_QUEUE_STREAM", f"ORION_QUEUE_{QUEUE_NAME.upper().replace('-', '_')}")
QTYPE = os.environ.get("ORION_QUEUE_TYPE", "work")
BASE_GROUP = os.environ.get("ORION_QUEUE_GROUP", f"{QUEUE_NAME}-workers")
REPLICA = os.environ.get("ORION_REPLICA_INDEX", "0")
# For topic queues every replica needs its own durable so JetStream tracks an
# independent cursor — that's what makes broadcast work. For work queues every
# replica shares the same durable so JetStream load-balances messages.
GROUP = BASE_GROUP if QTYPE == "work" else f"{BASE_GROUP}-r{REPLICA}"
LABEL = f"{QUEUE_NAME}#r{REPLICA}"


async def main() -> None:
    print(
        f"[{LABEL}] starting — type={QTYPE} subject={SUBJECT} stream={STREAM} group={GROUP}",
        flush=True,
    )
    nc = await nats.connect(NATS_URL)
    js = nc.jetstream()

    # Ensure the stream exists — `orion queue pub` does the same. Idempotent.
    try:
        await js.stream_info(STREAM)
    except NotFoundError:
        await js.add_stream(StreamConfig(name=STREAM, subjects=[SUBJECT]))
        print(f"[{LABEL}] created stream {STREAM}", flush=True)

    consumer = await js.pull_subscribe(
        subject=SUBJECT,
        durable=GROUP,
        stream=STREAM,
        config=ConsumerConfig(ack_policy=AckPolicy.EXPLICIT),
    )
    print(f"[{LABEL}] bound to durable={GROUP}", flush=True)

    while True:
        try:
            msgs = await consumer.fetch(batch=1, timeout=10)
        except asyncio.TimeoutError:
            continue
        for m in msgs:
            try:
                payload = m.data.decode("utf-8", errors="replace")
                try:
                    row = json.loads(payload)
                except json.JSONDecodeError:
                    row = {"_raw": payload}
                row.setdefault("_subject", m.subject)
                handle(row)
                await m.ack()
            except Exception as e:  # noqa: BLE001
                # nak and let JetStream redeliver; another worker can try.
                print(f"[{LABEL}] handler error: {e!r} — naking", file=sys.stderr, flush=True)
                try:
                    await m.nak()
                except Exception:  # noqa: BLE001
                    pass


if __name__ == "__main__":
    asyncio.run(main())
